ifacemonitor/update_filter.go

```diff
@@ -0,0 +1,199 @@
+	var timerC <-chan time.Time
+
+	type timestampedUpd struct {
+		ReadyAt time.Time
+		Update  interface{} // AddrUpdate or LinkUpdate
+	}
+
+	updatesByIfaceIdx := map[int][]timestampedUpd{}
+
+mainLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Info("FilterUpdates: Context expired, stopping")
+			return
+		case linkUpd := <-linkInC:
+			idx := int(linkUpd.Index)
+			if len(updatesByIfaceIdx[idx]) == 0 {
+				// We do see link updates as part of a flap but we've only seen them after the IP is
+				// flapped down.  Avoid delaying unrelated link updates.
+				logrus.Debug("FilterUpdates: link change with empty queue, short circuit.")
+				linkOutC <- linkUpd
+				continue
+			}
+			// Otherwise, we might have a flap in progress, delay the update.
+			updatesByIfaceIdx[idx] = append(updatesByIfaceIdx[idx],
+				timestampedUpd{
+					ReadyAt: u.Time.Now().Add(FlapDampingDelay),
+					Update:  linkUpd,
+				})
+		case addrUpd := <-addrInC:
+			idx := addrUpd.LinkIndex
+			oldUpds := updatesByIfaceIdx[idx]
+
+			var readyToSendTime time.Time
+			if addrUpd.NewAddr {
+				logrus.WithField("addr", addrUpd.LinkAddress).Debug("FilterUpdates: got address ADD")
+				if len(oldUpds) == 0 {
+					// This is an add for a new IP and there's nothing else in the queue for this interface.
+					// Short circuit.  We care about flaps where IPs are temporarily removed so no need to
+					// delay an add.
+					logrus.Debug("FilterUpdates: add with empty queue, short circuit.")
+					addrOutC <- addrUpd
+					continue
+				}
+
+				// Else, there's something else in the queue, need to process the queue...
+				logrus.Debug("FilterUpdates: add with non-empty queue.")
+				// We don't actually need to delay the add itself so we don't set any delay here.  It will
+				// still be queued up behind other updates.
+				readyToSendTime = u.Time.Now()
+			} else {
+				// Got a delete, it might be a flap so queue the update.
+				logrus.WithField("addr", addrUpd.LinkAddress).Debug("FilterUpdates: got address DEL")
+				readyToSendTime = u.Time.Now().Add(FlapDampingDelay)
+			}
+
+			// Coalesce updates for the same IP by squashing any previous updates for the same CIDR before
+			// we append this update to the queue.  We need to scan the whole queue because there may be
+			// updates for different IPs in flight.
+			upds := oldUpds[:0]
+			for _, upd := range oldUpds {
+				logrus.WithField("previous", upd).Debug("FilterUpdates: examining previous update.")
+				if oldAddrUpd, ok := upd.Update.(netlink.AddrUpdate); ok {
+					if ipNetsEqual(oldAddrUpd.LinkAddress, addrUpd.LinkAddress) {
+						// New update for the same IP, suppress the old update
+						logrus.WithField("address", oldAddrUpd.LinkAddress.String()).Debug(
+							"Received update for same IP within a short time, squashed the old update.")
+						continue
+					}
+				}
+				upds = append(upds, upd)
+			}
+			upds = append(upds, timestampedUpd{ReadyAt: readyToSendTime, Update: addrUpd})
```

Comment:
That was defensive. I was worried there might be a code path that keys a retry off the propagated update.  For example, if an interface flaps down and then back up we need to check that it's still in the right state because flapping down scrubs IPs and routes from the interface.

Reply:
OK, I see that there could equally be problems the other way round, so happy with this as is for now.